src/Corax/Querying/Matches/TermProviders/TermProvider.NumericRange.cs

```diff
@@ -59,56 +59,74 @@ false when typeof(TLow) == typeof(Range.Inclusive) && typeof(TVal) == typeof(Dou
             TVal startKey = _iterator.IsForward ? _low : _high;
             TVal finalKey = _iterator.IsForward ? _high : _low;
 
+            // Seek for startKey in the iterator
             _iterator.Seek(startKey);
-            if (_iterator.MoveNext(out TVal key, out _, out _) == false)
+            
+            // We get either
+            // currentKey equal to startKey
+            // currentKey equal to element succeeding startKey
+            // MoveNext returns false
+            if (_iterator.MoveNext(out TVal currentKey, out _, out _) == false)
             {
+                // When MoveNext returns false it means we jumped over all values stored inside a lookup tree
                 _isEmpty = true;
                 return;
             }
 
-            var skipFirst = _iterator.IsForward switch
-            {
-                true when typeof(TLow) == typeof(Range.Exclusive) && key.IsEqual(_low) => true,
-                false when typeof(THigh) == typeof(Range.Exclusive) && _high.CompareTo(key) <= 0 => true,
-                false when typeof(THigh) == typeof(Range.Inclusive) && _high.CompareTo(key) < 0 => true,
-                _ => false
-            };
-
+            // The first element may be equal to the startKey. For exclusive range start we want to skip it
+            var skipFirst = (_iterator.IsForward ? typeof(TLow) : typeof(THigh)) == typeof(Range.Exclusive) 
+                            && startKey.IsEqual(currentKey);
+            
+            // If we're supposed to skip first element, we rewrite the range to start from element succeeding startKey
             if (skipFirst)
             {
-                if (_iterator.MoveNext(out key, out _, out _) == false)
+                if (_iterator.MoveNext(out currentKey, out _, out _) == false)
                 {
+                    // We moved past queried range, nothing matches the query
                     _isEmpty = true;
                     return;
                 }
-                
-                if (_iterator.IsForward)
-                    _low = key;
-                else
-                    _high = key;
             }
             
+            // Update the range accordingly to the iterator option
+            if (_iterator.IsForward)
+                _low = currentKey;
+            else
+                _high = currentKey;
+            
             if (_skipRangeCheck)
-                return; // to the end
+                return;
             
-            //Now seek to the last key
+            // Seek for finalKey in the iterator
             _iterator.Seek(finalKey);
-            if (_iterator.MoveNext(out key, out _lastTermId, out var hasPreviousValue) == false)
+            
+            // We get either
+            // currentKey equal to finalKey
+            // currentKey equal to element succeeding finalKey
+            // MoveNext returns false
+            if (_iterator.MoveNext(out currentKey, out _lastTermId, out _) == false)
```

Comment:
Same here, explain what you are doing, why and what to expect. That you are going to do something when the iteration return is false is clear from the code, what it means for the algorithm is an entire different thing.

Reply:
Changed